demo-circuit/src/naive_threshold_sig_w_key_rotation/tests/next_cert.rs

```diff
@@ -0,0 +1,631 @@
+    );
+
+    let msg_to_sign = cert_to_msg(&withdrawal_certificate);
+    let wcert_signatures = create_signatures(
+        MAX_PKS,
+        &signing_keys_sks,
+        &validator_key_updates.signing_keys,
+        msg_to_sign,
+    );
+
+    let circuit_res = NaiveThresholdSignatureWKeyRotation::new(
+        validator_key_updates.clone(),
+        wcert_signatures,
+        withdrawal_certificate.clone(),
+        Some(prev_withdrawal_certificate.clone()),
+        THRESHOLD as u64,
+        genesis_validator_keys_tree_root,
+    );
+    assert!(circuit_res.is_ok());
+    let circuit = circuit_res.unwrap();
+
+    let debug_res = debug_circuit(circuit.clone());
+    assert!(debug_res.is_ok());
+
+    let old_genesis_validator_keys_tree_root = genesis_validator_keys_tree_root;
+    genesis_validator_keys_tree_root = validator_key_updates.get_upd_validators_keys_root().unwrap();
+    // Now try to use an outdated key
+    prev_withdrawal_certificate = withdrawal_certificate.clone();
+    validator_key_updates
+        .updated_signing_keys_sk_signatures[0] = NULL_CONST.null_sig;
+    validator_key_updates
+        .updated_signing_keys_mk_signatures[0] = NULL_CONST.null_sig;
+    validator_key_updates.signing_keys = validator_key_updates.updated_signing_keys.clone();
+
+    withdrawal_certificate = create_withdrawal_certificate();
+    withdrawal_certificate.custom_fields[0] = old_genesis_validator_keys_tree_root;
+    let msg_to_sign = cert_to_msg(&withdrawal_certificate);
+    let wcert_signatures = create_signatures(
+        MAX_PKS,
+        &signing_keys_sks,
+        &validator_key_updates.signing_keys,
+        msg_to_sign,
+    );
+
+    let circuit_res = NaiveThresholdSignatureWKeyRotation::new(
+        validator_key_updates.clone(),
+        wcert_signatures,
+        withdrawal_certificate.clone(),
+        Some(prev_withdrawal_certificate.clone()),
+        THRESHOLD as u64,
+        genesis_validator_keys_tree_root,
+    );
+    assert!(circuit_res.is_ok());
+    let circuit = circuit_res.unwrap();
+
+    let debug_res = debug_circuit(circuit.clone());
+    assert!(debug_res.is_ok());
+    let failing_constraint = debug_res.unwrap();
+    assert!(failing_constraint.is_some());
+}
+
+#[serial]
+#[test]
+fn bad_cert_hashes() {
+    const MAX_PKS: usize = VALIDATORS_SIZE;
+    const THRESHOLD: usize = 3;
+
+    let mut rng = thread_rng();
+    let (
+        signing_keys_sks,
+        _master_keys_sks,
+        genesis_validator_keys_tree_root,
+        withdrawal_certificate,
+        _wcert_signatures,
+        validator_key_updates,
+    ) = setup_certificate_data(MAX_PKS, THRESHOLD);
+
+    let mut prev_withdrawal_certificate = create_withdrawal_certificate();
+    prev_withdrawal_certificate.custom_fields[0] = genesis_validator_keys_tree_root;
+    prev_withdrawal_certificate.quality = THRESHOLD as u64;
+
+    let msg_to_sign = cert_to_msg(&withdrawal_certificate);
+    let wcert_signatures = create_signatures(
```

Comment:
It is unclear to me why it is needed to recompute the signatures on the withdrawal certificate